From: Ben Hutchings <ben@decadent.org.uk>
Date: Wed, 13 Apr 2016 21:48:06 +0100
Subject: fs: Add MODULE_SOFTDEP declarations for hard-coded crypto drivers
Bug-Debian: https://bugs.debian.org/819725

This helps initramfs builders and other tools to find the full
dependencies of a module.

Signed-off-by: Ben Hutchings <ben@decadent.org.uk>
---
 fs/btrfs/hash.c         | 2 ++
 fs/cifs/cifsencrypt.c   | 2 ++
 fs/cifs/link.c          | 2 ++
 fs/cifs/smb2transport.c | 2 ++
 fs/cifs/smbencrypt.c    | 2 ++
 fs/crypto/keyinfo.c     | 2 ++
 fs/ext4/crypto_key.c    | 2 ++
 fs/ext4/super.c         | 2 ++
 fs/f2fs/super.c         | 2 ++
 fs/jbd2/journal.c       | 2 ++
 fs/nfsd/nfs4recover.c   | 2 ++
 11 files changed, 22 insertions(+)

--- a/fs/btrfs/hash.c
+++ b/fs/btrfs/hash.c
@@ -13,6 +13,7 @@
 
 #include <crypto/hash.h>
 #include <linux/err.h>
+#include <linux/module.h>
 #include "hash.h"
 
 static struct crypto_shash *tfm;
@@ -44,3 +45,5 @@ u32 btrfs_crc32c(u32 crc, const void *ad
 
 	return *ctx;
 }
+
+MODULE_SOFTDEP("pre: crypto-crc32c");
--- a/fs/cifs/cifsencrypt.c
+++ b/fs/cifs/cifsencrypt.c
@@ -875,3 +875,5 @@ cifs_crypto_shash_release(struct TCP_Ser
 	kfree(server->secmech.sdescmd5);
 	server->secmech.sdescmd5 = NULL;
 }
+
+MODULE_SOFTDEP("pre: crypto-arc4 crypto-ecb crypto-hmac crypto-md5");
--- a/fs/cifs/link.c
+++ b/fs/cifs/link.c
@@ -746,3 +746,5 @@ symlink_exit:
 	free_xid(xid);
 	return rc;
 }
+
+MODULE_SOFTDEP("pre: crypto-md5");
--- a/fs/cifs/smb2transport.c
+++ b/fs/cifs/smb2transport.c
@@ -700,3 +700,5 @@ smb2_setup_async_request(struct TCP_Serv
 
 	return mid;
 }
+
+MODULE_SOFTDEP("pre: crypto-aes crypto-cmac crypto-hmac crypto-sha256");
--- a/fs/cifs/smbencrypt.c
+++ b/fs/cifs/smbencrypt.c
@@ -259,3 +259,5 @@ SMBNTencrypt(unsigned char *passwd, unsi
 	rc = E_P24(p21, c8, p24);
 	return rc;
 }
+
+MODULE_SOFTDEP("pre: crypto-des crypto-ecb crypto-md4");
--- a/fs/crypto/keyinfo.c
+++ b/fs/crypto/keyinfo.c
@@ -10,6 +10,7 @@
 
 #include <keys/encrypted-type.h>
 #include <keys/user-type.h>
+#include <linux/module.h>
 #include <linux/random.h>
 #include <linux/scatterlist.h>
 #include <uapi/linux/keyctl.h>
@@ -302,3 +303,5 @@ int fscrypt_get_encryption_info(struct i
 	return 0;
 }
 EXPORT_SYMBOL(fscrypt_get_encryption_info);
+
+MODULE_SOFTDEP("pre: crypto-aes crypto-ecb");
--- a/fs/ext4/crypto_key.c
+++ b/fs/ext4/crypto_key.c
@@ -11,6 +11,7 @@
 #include <crypto/skcipher.h>
 #include <keys/encrypted-type.h>
 #include <keys/user-type.h>
+#include <linux/module.h>
 #include <linux/random.h>
 #include <linux/scatterlist.h>
 #include <uapi/linux/keyctl.h>
@@ -272,3 +273,5 @@ int ext4_has_encryption_key(struct inode
 
 	return (ei->i_crypt_info != NULL);
 }
+
+MODULE_SOFTDEP("pre: crypto-aes crypto-ecb");
--- a/fs/ext4/super.c
+++ b/fs/ext4/super.c
@@ -5455,6 +5455,8 @@ static void __exit ext4_exit_fs(void)
 	ext4_exit_es();
 }
 
+MODULE_SOFTDEP("pre: crypto-crc32c");
+
 MODULE_AUTHOR("Remy Card, Stephen Tweedie, Andrew Morton, Andreas Dilger, Theodore Ts'o and others");
 MODULE_DESCRIPTION("Fourth Extended Filesystem");
 MODULE_LICENSE("GPL");
--- a/fs/f2fs/super.c
+++ b/fs/f2fs/super.c
@@ -1749,6 +1749,8 @@ static void __exit exit_f2fs_fs(void)
 	f2fs_destroy_trace_ios();
 }
 
+MODULE_SOFTDEP("pre: crypto-crc32c");
+
 module_init(init_f2fs_fs)
 module_exit(exit_f2fs_fs)
 
--- a/fs/jbd2/journal.c
+++ b/fs/jbd2/journal.c
@@ -2715,6 +2715,8 @@ static void __exit journal_exit(void)
 	jbd2_journal_destroy_caches();
 }
 
+MODULE_SOFTDEP("pre: crypto-crc32c");
+
 MODULE_LICENSE("GPL");
 module_init(journal_init);
 module_exit(journal_exit);
--- a/fs/nfsd/nfs4recover.c
+++ b/fs/nfsd/nfs4recover.c
@@ -1560,3 +1560,5 @@ unregister_cld_notifier(void)
 {
 	rpc_pipefs_notifier_unregister(&nfsd4_cld_block);
 }
+
+MODULE_SOFTDEP("pre: crypto-md5");
